### ISSUE TITLE
RMSMappingQuality explodes when there are 0 reads

### ISSUE THREAD

**Issue by droazen** [broadinstitute/gatk#2658]
@kcibul reports getting the following error while running `GenotypeGVCFs`:

```
New error coming out of GenotypeGVCFs when running 1000 shards:

[11:20] 
***********************************************************************

A USER ERROR has occurred: Bad input: Cannot calculate Root Mean Square Mapping Quality if there are 0 or less reads.
Number of reads recorded as :0
In VariantContext: [VC UG_call @ chr1:143249103 Q38.86 of type=SNP alleles=[C*, T] attr={DP=14, ExcessHet=3.0102999210357666, MLEAC=[2], MLEAF=[0.25], RAW_MQ=0.0} GT=[[09C81377 C*/C* GQ 6 DP 5 PL 0,6,90 {MIN_DP=4}],[09C83237 C*/C* GQ 3 DP 6 PL 0,3,45 {MIN_DP=6}],[09C97255 ./. DP 0 PL 0,0,0 {MIN_DP=0}],[09C98651 ./. DP 0 PL 0,0,0 {MIN_DP=0}],[09C99383 T/T GQ 3 PL 45,3,0 {PGT=0|1, PID=143249097_C_T, SB=[0, 0, 0, 0]}],[09C99677 ./. DP 0 PL 0,0,0 {MIN_DP=0}],[10C100868 C*/C* GQ 3 DP 7 PL 0,3,45 {MIN_DP=4}],[10C101312 ./. DP 0 PL 0,0,0 {MIN_DP=0}],[10C102545 ./. DP 1 PL 0,0,0 {MIN_DP=0}],[10C102782 ./. DP 0 PL 0,0,0 {MIN_DP=0}]]

[11:21] 
```

I think this might be a bug, or at least a case of over-aggressive error checking. We have an extra check in our version of `RMSMappingQuality` that is not present in the GATK3 version:

```
        if (numOfReads <= 0){
            throw new UserException.BadInput("Cannot calculate Root Mean Square Mapping Quality if there are 0 or less reads." +
                                                     "\nNumber of reads recorded as :" +numOfReads +
                                                     "\nIn VariantContext: "+ vc.toStringDecodeGenotypes());
        }
```

We should match the GATK3 behavior in this case, unless we can **prove** (and not merely infer) that GATK3 also explodes.

@lbergelson Since you are git blamed here, can you give your thoughts on this?

**Comment by droazen** [comment]
Would also be good if annotation expert @ldgauthier could chime in with her thoughts on this as well.

**Comment by ldgauthier** [comment]
Two weird things about the GTs -- sample 09C99383 is homVar but has no reported DP and sample 10C102545 is a DP 1 no-call (probably not the issue)

@kcibul are the gVCFs and TileDB used as input still around?  Something odd happened in at least one place upstream of this error.

Also, should be "0 or fewer reads" :-P

**Comment by lbergelson** [comment]
@droazen @kcibul Would it be possible to dump the combinedGvcf from the merged tiledb here so we could try running that line through gatk3?  

I'm very certain that if that method exits with 0 in gatk3 or gatk4 it will immediately result in a divide by 0 exception.  I can't be 100% sure that this case would exit with 0 in gatk3 without actually testing it though.  

If we were to assume that the result of this is 0 in both gatk3 and 4, what would the right thing to do here be?  Output 0 for the annotation?

**Comment by ldgauthier** [comment]
I'd like a gVCF snippet for sample 09C99383 if that's possible.  If we assume that genomicDB is faithfully merging gVCFs, then there was something wrong going in.

**Comment by kcibul** [comment]
On it --

**Comment by kcibul** [comment]
@ldgauthier the GVCF for that sample is at gs://broad-jg-dev-storage/temp/09C99383.91c5a812-70c5-4526-a3a2-3e99b9cf08fb.g.vcf.gz

**Comment by kcibul** [comment]
... and the list of the 10 GVCFs that were used are at:
 gs://broad-jg-dev-storage/jg_scale_testing/hg38_scale_test_10.gvcf_list

**Comment by ldgauthier** [comment]
I'm not sure I have access to that bucket -- which project is it associated with?

**Comment by lbergelson** [comment]
I also can't access those files.  Who can grant access to them?

**Comment by lbergelson** [comment]
@ldgauthier A copy of the generated genomics db and the combined gvcf extracted from it with SelectVariants is here: gs://broad-dsde-methods/kcibul/bug_for_louis

**Comment by ldgauthier** [comment]
It doesn't explode in GATK3 -- I had to drop -stand_call_conf to 10 to get it to output, then I got MQ=NaN.

The NaN/explosion is my fault -- pulling the DP from the format field seemed like such a good idea at the time, but it's just causing problems.  I had a ticket open for it, but the particular test case got fixed in another way.  I'll open another ticket.

**Comment by ldgauthier** [comment]
@kcibul I'd still like the gVCFs for debugging and as test data for https://github.com/broadinstitute/gatk/issues/2668

**Comment by droazen** [comment]
@ldgauthier We've decided to replicate the GATK3 behavior in GATK4 for now and just generate NaNs. This is preferable to an explosion, and would unblock @kcibul while we await the definitive fix you've proposed in https://github.com/broadinstitute/gatk/issues/2668. We should also still keep this ticket open to try to get to the bottom of the weirdness in the GTs you mentioned above, of course (eg., "sample 09C99383 is homVar but has no reported DP").

**Comment by ldgauthier** [comment]
Yep, that's what I expected.  VQSR will smooth over the NaNs.

**Comment by droazen** [comment]
Re-opening this issue -- we shouldn't close it until we've examined the GVCFs in question, as there might be deeper problems here.

**Comment by ldgauthier** [comment]
I think the gvcfs are okay - I got confused by the genotype toString even
though Louis warned me that it's terrible.

---------------------------------------
Laura Doyle Gauthier, PhD
Computational Biologist
Data Sciences and Data Engineering
Broad Institute of MIT and Harvard
gauthier@broadinstitute.org

**Comment by droazen** [comment]
Closing -- this is done!
